### ISSUE TITLE
Integrate darkrho/scrapy-inline-requests

### ISSUE THREAD

**Issue by Curita** [scrapy/scrapy#1144]
I think we haven't created a ticket for this yet, we discussed integrating https://github.com/darkrho/scrapy-inline-requests a couple of weeks ago with @pablohoffman, @kmike and @nramirezuy.

I'm copying the whole discussion here if anyone else wants to join in:

From Pablo:
>I've heard many people using (and speaking good things about) inline requests recently. Would you consider it a feature to include in 1.0?

From Mikhail:
>I'd like to have something like inline-requests builtin; it even was one of the draft ideas for 2014 GSoC :) +1 to add it to Scrapy 1.0 if we have time for that.

>Past me wrote the following in one of the emails:

>> It needs a bit more thought before becoming a part of Scrapy: there are no tests, I'm not fan of how callbacks are handled, and downsides of the 'yield' approach should be clearly documented - e.g. state inside the callback lives longer, and it could lead to increased memory usage; it is also unclear how does it work with on-disk queues. There are also some useful features that present in other alike libraries (e.g. adisp) but not in scrapy-inline-requests - for example, waiting for several requests to be executed in parallel: syntax could be resp1, resp2 = yield Request(url1), Request(url2)

From Nicolás:
>I like the idea behind inline requests, but not the API of it. It kinda doesn't fit the callback approach since it doesn't work with  a callback and you have to manage several requests with in a callback. 

>I would prefer to see something like:
>def callback:
>    return chain_requests(request1, request2, request3)

>and the callbacks handled normally.

From Mikhail:
>Nicolás: I think the point of inline-requests is to allow writing code without callbacks and handle several related requests in a single function :) It is a common trick to "linearize" callbacks into a generator.

>Callbacks + CPython reference counting (no PyPy) provide a nice approach to resource deallocation: if a variable is not referenced from outside then it is deallocated as soon as the callback exits, without invoking garbage collector. With generators if user writes "response1 = yield ...; response2 = yield ..." then these responses are kept alive, possibly for long. Even with "response = yield ...; response = yield ..." response is kept in memory longer than needed (if I'm not mistaken, until the second request finishes). One can write "del response", but it'd be nice to have some clever solution for that.

From Julia:
>I wouldn't promote it as the preferred way for dealing with requests/responses because of the already mention issues. It's not as flexible as using explicit callbacks (we should document that yielding a request with `callback` not being `None` breaks the chain btw) and it's hacky, debugging it is kind of hard.

>Still, it's a really good helper for its primary use-case of downloading some additional page and handling errors (as opposed to using errbacks or downloading the page with another library) so I'd also like to include it in Scrapy.

NOTE: I didn't mean that I breaks the chain as in raising an exception, just that it won't wait for yielded requests if they have callbacks.

From Pablo:
>I think we're pretty much in agreement that it would be a nice feature for 1.0 (well, if we're not gonna have python 3 ... :). It needs to go with good documentation (explaining the downsides), tests and better error check (raising exception if it's used with a request having a callback).

>Shall we make a ticket for this?. I think there's already enough content in this thread for one :).

/cc @darkrho

**Comment by kmike** [comment]
An idea: we can try solving memory issues by setting higher priority for inline requests. This way Scrapy will prefer to finish current requests in `inline_requests` method first, before starting more such request chains.

**Comment by darkrho** [comment]
>  it's hacky, debugging it is kind of hard

I totally agree with that. :)

**Comment by kmike** [comment]
### Resource deallocation

This is kind of important because even in this code:

```python
resp = yield Request(url1)
resp = yield Request(url2)
```

the first `resp` value lives in memory until Scrapy returns a response for the second Request. It means users must write this:

```python
resp = yield Request(url1)
del resp
resp = yield Request(url2)
```

I don't like this; even if documented, it is easy to miss and obscures spider logic. One way to fix it is to ask users to write code like this:

```python
data1 = process1(yield Request(url1))
data2 = process2(yield Request(url2))
```
but I'm not sure it is that better than callbacks. 

Alternatively, we can provide a context manager which will kill all heavy Response attributes (at least delete self.body and self.selector, and maybe even kill all alive selectors for this response) on exit and so won't let users shoot themselves in the foot:

```python
with (yield fetch(url1)) as resp:
    ...
with (yield fetch(url2)) as resp:
    ...
```

### Disk queues

It is not possible to resume generator from the middle, at least without hacking frame objects using a C extension. This means the best we can get is 'restart the whole coroutine if it is not processed in full'. But this is also tricky because we must somehow 'acknowledge' a request when coroutine exits, not when request with a coroutine callback finishes.

### PEP 492 support

There are some advantages in using [PEP 492](https://www.python.org/dev/peps/pep-0492/) coroutines instead of generators wrapped in decorators:

* it is a native feature;
* better debugging (pdb support, better tracebacks);
* nicer syntax for `async with` (braces are not needed);
* `async for` support which can be useful for e.g. start_requests or `idle` handlers.
* missing `await` calls are easier to notice because of [PYTHONASYNCIODEBUG](https://www.python.org/dev/peps/pep-0492/#debugging-features)

The problem is that if we want to support `async def`, `await`, `async for` and `async with` then we can't use existing Scrapy API because `yield` is not allowed in coroutines (at least in Python 3.5). It means we'll have to change the API to something like this:

```python
class MySpider(scrapy.Spider):
    # ...
    async def parse(self, response):
        for link in response.links():
            self.send_request(link.url, self.parse2)  # async request 
        async with self.fetch(url1) as resp:  # wait for the request
            data = resp.css('div.foo::text').extract_first()
            self.store_item({'url': resp.url, 'data': data})
```

An alternative which won't require us to make Spider more 'fat' is to use Grab-like API:

```python
async def parse(crawler):  
    for link in crawler.response.links():
        crawler.async_request(link.url, parse2) 
    async with crawler.get(url1) as resp:  # wait for the request
        data = resp.css('div.foo::text').extract_first()
        crawler.store_item({'url': resp.url, 'data': data})
```

Either way, `async def` and regular callbacks can be mixed (scrapy can check if a callback is a coroutine and act accordingly), so the change can be backwards-compatible.

### inline_requests API

In inline_requests Request object is used both to send requests we should and shouldn't wait for. I think they should be separated more explicitly (like crawler.async_request and crawler.get/post/.. in the last PEP492 example).

**Comment by kmike** [comment]
Obvious observation: an alternative to 
```
resp = yield Request(url1)
del resp
resp = yield Request(url2)
```
is

```
resp = yield Request(url1)
del resp; resp = yield Request(url2)
```

**Comment by stav** [comment]
I think this thread is interesting because it brings up what Mikhail pointed out PEP-492, that built-in support for async in Python 3.5 uses the `await` expression which is like `yield from` from previous versions which scrapy-inline-requests is using to suspend execution _inline_.  So this is what Python is outlining for its future of async.  And this is, I think, the real sub-surface topic that is going on in this thread:

_(When) Do we want to change the Scrapy API from callbacks to linline?_

**Comment by kmike** [comment]
Yeah, it is not specifically about inline_requests. I think we shouldn't *change* Scrapy API from callbacks to coroutines, we should provide *additional* coroutine-based API: while looking nice couroutines have their downsides.

**Comment by patiences** [comment]
Hi there!! I'm a little confused about the proposed API above: 

>
>```
>class MySpider(scrapy.Spider):
>    # ...
>    async def parse(self, response):
>        for link in response.links():
>            self.send_request(link.url, self.parse2)  # async request 
>        async with self.fetch(url1) as resp:  # wait for the request
>            data = resp.css('div.foo::text').extract_first()
>            self.store_item({'url': resp.url, 'data': data})
>```

Looks like this piece of code:
1. Gets all the links from the downloaded `response` and sends an async request for each one, with `parse2` as the callback 
2. Saves the data retrieved from `url1` using `store_item` (because `yield` expressions cannot exist in async functions)

Are you proposing to add new methods `links`, `send_request`, `fetch`, `store_item`? 
What exactly is `self.fetch(url1)`? 
`parse2` could be a coroutine as well if there are more links to follow, right? 

> I think we shouldn't change Scrapy API from callbacks to coroutines
Out of curiosity, why not? Would you mind elaborating on the downsides that you mentioned?

**Comment by kmike** [comment]
Hey @patiences!

As async generators are implemented in Python 3.6, I think that's fine to use them instead. With async generators this code would look like this:

```py
class MySpider(scrapy.Spider):
   # ...
   async def parse(self, response):
       for link in response.links():
           yield scrapy.Request(link.url, self.parse2)  # async request 
       async with self.fetch(url1) as resp:  # wait for the request
           data = resp.css('div.foo::text').extract_first()
           yield {'url': resp.url, 'data': data}
```

or, if we ditch `async with`:

```py
class MySpider(scrapy.Spider):
   # ...
   async def parse(self, response):
       for link in response.links():
           yield scrapy.Request(link.url, self.parse2)  # async request 

       # wait for the request, i.e. handle it in the same callback
       resp = await self.fetch(url1) 
       # even better, though I'm not sure if this is possible:      
       # resp = await scrapy.Request(url1)
       data = resp.css('div.foo::text').extract_first()
       yield {'url': resp.url, 'data': data}
```


To make it clear: I'm putting these examples for discussion, it is not a proposal yet.

Regular Scrapy callbacks yield Request objects with calbacks; Scrapy calls a callback when response is ready. The idea is to provide a way to handle responses in the same callback; this is helpful when you're scraping data from several related pages to create a single output item (e.g. you're scraping products, and shipping price is only available when you click "add to cart").

> Out of curiosity, why not? Would you mind elaborating on the downsides that you mentioned?

With the current Scrapy API when you send a request, a request can be put to a disk queue; this way you can survive restarts, and also save RAM. Also, as there is a new callback handling each response, there is only a little responses kept in memory at the same time, it also saves RAM. So I think async API should be optional, for cases where it makes thinkg simpler. It is easier to shoot yourself in a foot and make your spider inefficient with these async def functions. Also, keep in mind we'd love to keep backwards compatibility, so "change" could be a wrong word.

**Comment by kmike** [comment]
FTR, `await scrapy.Request(url1)` syntax is hard, because in this case scrapy.Request needs to schedule itself - it needs a reference to Crawler, which it currently doesn't have by design. As I recall, you may need something similar to https://bitbucket.org/hipchat/txlocal, or maybe something similar to https://www.python.org/dev/peps/pep-0567/ (implemented in upcoming Python 3.7). This is tricky.

**Comment by patiences** [comment]
@kmike thanks for the code samples - that helps a lot. :-) 

> The idea is to provide a way to handle responses in the same callback

Ah ha - got it, thanks. 

Do you lean towards either option? I might prefer to have `await` rather than `async with` just because it feels more idiomatic to see them together. As for `resp = await self.fetch(url1)` vs `resp = await scrapy.Request(url1)`, I agree that the latter is more intuitive but would using a thread-like object to self-schedule requests contradict the Engine/Scheduler architecture? 

> So I think async API should be optional, for cases where it makes thinkg simpler. It is easier to shoot yourself in a foot and make your spider inefficient with these async def functions. 

Agree completely. Maybe it's good thing then that the API will have to change substantially, since it forces people to think about whether or not they really need that behavior. :-)

**Comment by yashrsharma44** [comment]
Thanks for the insights @kmike .
There were few queries in the async def parse -
```python
class MySpider(scrapy.Spider):
   # ...
   async def parse(self, response):
       for link in response.links():
           yield scrapy.Request(link.url, self.parse2)  # async request 
       async with self.fetch(url1) as resp:  # wait for the request
           data = resp.css('div.foo::text').extract_first()
           yield {'url': resp.url, 'data': data}
```
Regarding parse2, can't we call parse recursively ?
Also can't we use the yielded value of scrapy.Request in the asynchronous context manager ?

**Comment by kmike** [comment]
> Do you lean towards either option? I might prefer to have await rather than async with just because it feels more idiomatic to see them together.

@patiences the difference is that with a context manager we clean up resources for the user - spiders will use less memory. See https://github.com/scrapy/scrapy/issues/1144#issuecomment-141843616 for motivation. Currently I'm leaning towards just `await`. Maybe we should provide both.

> As for resp = await self.fetch(url1) vs resp = await scrapy.Request(url1), I agree that the latter is more intuitive but would using a thread-like object to self-schedule requests contradict the Engine/Scheduler architecture?

The idea is to have a coroutine-specific context where Scrapy can put a reference to the Crawler. See examples at https://github.com/Skyscanner/aiotask-context. In Python 3.7 they're built-in and supported natively in asyncio (see https://docs.python.org/dev/library/contextvars.html#asyncio-support and https://www.python.org/dev/peps/pep-0567/). There should be a way to do the same in Twisted. Or a way to run Twisted on asyncio even loop. Request can then get Crawler object from the context, and ask Crawler to schedule itself. So Engine/Scheduler/.. architecture stays the same, we just provide a way for Request to access Crawler when it is necessary.

> Regarding parse2, can't we call parse recursively ?

@yashrsharma44 I'm not sure I understood your question properly. We can use self.parse instead of sef.parse2. Why are you asking?

> Also can't we use the yielded value of scrapy.Request in the asynchronous context manager ?

Do you mean this?

```py
    async with scrapy.Request(url1) as resp:  # wait for the request
       data = resp.css('div.foo::text').extract_first()
       yield {'url': resp.url, 'data': data}
```

Hm, it could work. I like that we're not adding new Spider methods in this case. Though I don't quite like how it reads: "with request as response" - as if request and response are the same thing, and we're using request as response. "with fetch(request) as response" reads better, as we're more explicit that response is a result of fetching a request.

**Comment by yashrsharma44** [comment]
Thanks @kmike .
>@yashrsharma44 I'm not sure I understood your question properly. We can use self.parse instead of sef.parse2. Why are you asking?

You interpreted it correctly. 
 One final query : How are we looking to maintain backwards compatibility ? Are we going to generate this as a separate API, from scrapy ? And then figure out how to backport it ?

**Comment by kmike** [comment]
> How are we looking to maintain backwards compatibility ? Are we going to generate this as a separate API, from scrapy ? And then figure out how to backport it ?

We can check if a callback is a regular method or `async def` method; if it is a regular function, do exactly as we were doing before, for `async def ` functions support all this stuff. I haven't checked how to do this though, but that's an idea.

**Comment by ThobaniMwandla** [comment]
I'm quietly confused about this line shown below, I cant even see the fetch method and its url argument, please help:-
             async with self.fetch(url1) as resp:  # wait for the request

**Comment by wRAR** [comment]
As we are almost ready to merge the feature that allows `await` in callbacks (#4978), I've decided to check what can now be done to implement things discussed here.

So far the idea is to schedule the request via some API available in the callback and receive either a response or an exception (raised or returned) right there, pausing the callback but not the spider until it's available. As we support `await` the simplest API would be:

```
try:
    resp = await req
except:
    self.logger.error()
```

This also almost matches the inline-requests API, only replacing yield with await and removing the confusion between yielding a request normally and yielding it to get a response right here.

Implementing this simple API should be possible by adding an `__await__()` to the Request class which will do actual scheduling and return the response/raise an exception when the engine returns the result. It will most likely need `crawler` and `spider` which can be provided via contextvars (or passed directly but that's an inconvenient API). The implementation of this `Request.__await__()` can be trivial.

My first implementation calls `engine.crawl()`, with a simple change in `Scraper.call_spider()` that disables calling `.callback`/`.errback` and instead fires a special Deferred which is awaited in `Request.__await__()`. This seems to work with some caveats, but raises many questions (more on this later). An even more straightforward implementation just calls `engine.download()` and awaits on its return value. At this point it became clear to me than one can just call `resp = await engine.download()` in a callback, just as we already propose to do in our pipeline example, without any changes to the Scrapy code, which hardly a good API but something that can be used right after adding the async callback support. Anyway, while this is simple and straightforward, it skips spider middlewares so maybe it's not the best idea.

**Comment by wRAR** [comment]
While we didn't discuss the current implementation ideas much, we already found multiple questions:

* Should an exception be raised or returned in some form? Looks like raising is better for the API user.
* How should an awaited request be serialized, put into a disk queue, saved with the spider state to resume the spider from it, etc.? Unlike a normal detached request that only needs to know what callback to call with what cb_kwargs when the response is available, an executed and paused callback that awaits for a request result would need to be serialized as a whole, which doesn't seem reasonable.
* How is the additional memory usage problem solved?
* How should a request object with a set callback and/or errback be handled?
* Should a priority of an awaited request be raised by Scrapy or not?

Some of these problems can be answered by sending the request directly to the downloader instead of scheduling it, though this leads to some other questions.

Also, some of these problems may affect inline-requests too, but a list of inline-requests limitations is quite long and we want to have something better as a replacement.

**Comment by wRAR** [comment]
It also may be useful to consider if we need anything else in addition to a linear `await Request()`. In my experience, inline-requests is mostly used in the following scenarios:

1. There is too much state to pass around in `meta` cleanly, so a request, or several sequential ones, are made in the same callback. This is straightforward.
2. A bunch of subpages contain data that needs to be gathered into a single item together with data from the main page. While this can be trivially implemented with a loop, subpage requests can also be submitted in parallel as they are usually independent. Depending on the processing logic, the processing could be even started when the first request returns while waiting for the others. It may be possible to implement this with async generators, `asyncio.gather`/`asyncio.as_completed` etc., or it may require some special Scrapy support. See also #2600.

**Comment by kmike** [comment]
Nice summary, thanks @wRAR!

**Comment by kmike** [comment]
Some thoughts on the questions: 

> Should an exception be raised or returned in some form? Looks like raising is better for the API user.

+1 to raise an exception.

> How should an awaited request be serialized, put into a disk queue, saved with the spider state to resume the spider from it, etc.? Unlike a normal detached request that only needs to know what callback to call with what cb_kwargs when the response is available, an executed and paused callback that awaits for a request result would need to be serialized as a whole, which doesn't seem reasonable.

It doesn't seem to be possible to serialize and deserialize these requests like other requests. To restore spider state from the disk queue you'd need all the local variables in the callback, generator state if a callback is a generator, etc., which looks impossible, or very brittle and hacky if an implementation is possible.

My vote is to not put these requests to scheduler, and send them to downloader (with some caveats). 
An alternative is to force in-memory queue for such requests (with a different set of caveats).

> How is the additional memory usage problem solved?

If these requests don't go through Scheduler, they're not subject to usual prioritization, which means they're going to be processed faster, and won't block callbacks forever.

I think an issue with responses hanging around can be solved separately. `del response` workaround can be ok for now.

> How should a request object with a set callback and/or errback be handled?

Raise an exception if a callback is set. 
For errback it is a bit more nuanced - e.g. we could say that if errback is present, it is called instead of having an exception raised. But it is not clear what to return in this case (maybe None?), and what's the use case. I think we can raise an exception if errback is set for now; this can be changed later if there is a good proposal.

> Should a priority of an awaited request be raised by Scrapy or not?

This is solved on itself if request bypasses the Scheduler. I recall we don't repect priorities in downloader at the moment. 

If a request is sent to Scheduler, raising a priority won't do much - we don't know what are the priorities of other request, and queues may be implemented in a way that priorities only work within downloader slot, not globally. So by raising a priority we can't really prevent a request from being in a queue for a long time.

----

Regarding caveats on not sending requests through Scheduler. As @wRAR said, probably the biggest one is not having spider middlewares running. I've checked built-in middlewares, and some of them look important: 

* depth: we may not want depth limit to be working here, though this is up to discussion
* httperror: it'd be nice to have an exception raised for non-200 responses, so it'd be good to have it
* offsite - as discussed elsewhere, it probably needs to be a downloder middleware as well. Good to have.
* referer - good to have
* urllength - good to have

Most built-in spider middlewares make sense for inline requests; only depth middleware is unclear, but it is fine to have it working as well.

So, maybe we should have some very specific semantics for these requests: they go through all the middlewares, but skip the Scheduler?

**Comment by Gallaecio** [comment]
> My vote is to not put these requests to scheduler, and send them to downloader (with some caveats).
> An alternative is to force in-memory queue for such requests (with a different set of caveats).

At the moment I feel more inclined to having these requests also go through the scheduler, but do not allow to serialize them. I would rather not skip the scheduler for any request, at least not by default. Memory buildup should not be a concern, as [the default scheduler prioritizes unserializable requests](https://github.com/scrapy/scrapy/blob/ddea6b7bfa38bf5402d78350ab61e2c827ca49b5/scrapy/core/scheduler.py#L259). I’m not sure at the moment of which other caveats we need to be concerned about.

+1 to everything else.

**Comment by kmike** [comment]
A good point about prioritizing in-memory request @Gallaecio, I think you're right. I don't see other caveats here. Maybe we'd need to make sure that such requests are not seen as an errors from the logging/stats point of view (scrapy non-serializable requests), but that's minor.

**Comment by kmike** [comment]
One more thing: it would be cool to have await-based shortcuts for headers_received and bytes_received signals. Use cases: a different way to cancel a request; streaming of the responses.

**Comment by Gallaecio** [comment]
I wonder if [scrapy-poet‘s implementation of web-poet’s additional requests](https://github.com/scrapinghub/scrapy-poet/pull/62) could help here. It does not send requests through the scheduler, though.

**Comment by AdrianAtZyte** [comment]
I wonder if we really need to implement something that will go through spider middlewares. If you ignored that, `crawler.engine.download_async()` seems to address the original ask. [We document it as the inline requests approach now](https://docs.scrapy.org/en/latest/topics/coroutines.html#inline-requests).

**Comment by wRAR** [comment]
> I wonder if we really need to implement something that will go through spider middlewares. 

I still don't know, yeah.

>  If you ignored that, crawler.engine.download_async() seems to address the original ask. 

Yeah, it does, though making a wrapper that allows `await Request()` is nice (but 100% not *necessary* IIRC).
